### PR TITLE
Mongodb PCA service

### PR DESCRIPTION
**Pull request by Aminelouati**
In this PR, we:
1/ Add a new MongoDB PCA service composed of three states (Started, Stopped and Finished)
2/ Add some workflow examples to the database-services bucket showing how to use this service from the client-side
3/ Improve the MySQL and MongoDB workflows in data-connectors bucket (credential mechanism) to comply with the added service.
4/ Improve the output of some existing workflows.

### CODE REVIEW

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 1]
```diff
@@ -133,44 +134,52 @@ def write_mongo(db, collection, inputs, host, port, username, password, no_id=Tr
...
-    MONGODB_PASSWORD=credentials.get("MONGODB_PASSWORD")
-else:
-    print("You first need to add third-party credentials for the database in the scheduler-portal.")
+    print("The MONGODB_COLLECTION variable is not provided by the user.")
     sys.exit(1)
+if variables.get("MONGODB_USER"):
+    USER = variables.get("MONGODB_USER")
+    # This key is used for getting the password from 3rd party credentials.
+    MONGODB_PASSWORD_KEY = "mongodb://" + USER + "@" + HOSTNAME + ":" + str(PORT)
+    if credentials.get(MONGODB_PASSWORD_KEY):
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: Duplicated line ?

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 2]
```diff
@@ -133,44 +134,52 @@ def write_mongo(db, collection, inputs, host, port, username, password, no_id=Tr
...
+    if credentials.get(MONGODB_PASSWORD_KEY):
+        PASSWORD=credentials.get(MONGODB_PASSWORD_KEY)
+    else:
+        print("The MONGODB_USER and MONGODB_PASSWORD are used in junction. Please add your mongodb password to 3rd-party credentials in the scheduler-portal under the key :\"" + MONGODB_URL_KEY + "\"")
+        sys.exit(1)
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: Replace l.165- l.168 by

PASSWORD=credentials.get(MONGODB_PASSWORD_KEY)
if not PASSWORD
      print("......

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 3]
```diff
@@ -237,40 +246,47 @@ import json
...
 
-
 PORT = 27017
 MODE= 'QUERY'
+MONGODB_OUTPUT=None
+USER=""
+PASSWORD=""
+MONGODB_URL_KEY = "mongodb://<username>@<hostname>:<port>"
 
 if variables.get("MONGODB_HOSTNAME"):
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 4]
```diff
@@ -237,40 +246,47 @@ import json
...
     HOSTNAME = variables.get("MONGODB_HOSTNAME")
 else:
-    print("MONGODB_HOSTNAME is not defined by the user.")
+    print("The MONGODB_HOSTNAME variable is not provided by the user.")
     sys.exit(1)
 if variables.get("MONGODB_PORT"):
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 5]
```diff
@@ -237,40 +246,47 @@ import json
...
     PORT = int(variables.get("MONGODB_PORT"))
 else:
-    print("MONGODB_PORT is not defined by the user. Using the default value:"+PORT)
+    print("The MONGODB_PORT variable is not provided by the user. Using the default value:"+PORT)
 if variables.get("MONGODB_DATABASE"):
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MongoDB.xml (hunk 2 above)
Comment (Aminelouati) [reply DataConnectors/resources/catalog/MongoDB.xml]: it is handled in the else bloc

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 6]
```diff
@@ -237,40 +246,47 @@ import json
...
     DATABASE = variables.get("MONGODB_DATABASE")
 else:
-    print("MONGODB_DATABASE is not defined by the user.")
+    print("The MONGODB_DATABASE variable is not provided by the user.")
     sys.exit(1)
 if variables.get("MONGODB_COLLECTION"):
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 7]
```diff
@@ -237,40 +246,47 @@ import json
...
-    sys.exit(1)
-if credentials.get("MONGODB_USERNAME") is not None and credentials.get("MONGODB_PASSWORD") is not None:
-    MONGODB_USERNAME = credentials.get("MONGODB_USERNAME")
-    MONGODB_PASSWORD=credentials.get("MONGODB_PASSWORD")
-else:
-    print("You first need to add third-party credentials for the database in the scheduler-portal.")
+    print("The MONGODB_COLLECTION variable is not provided by the user.")
     sys.exit(1)
+if variables.get("MONGODB_USER"):
+    USER = variables.get("MONGODB_USER")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: USER = variables.get("MONGODB_USER")
if USER :
    ...

File: DataConnectors/resources/catalog/MongoDB.xml [hunk 8]
```diff
@@ -237,40 +246,47 @@ import json
...
+    # This key is used for getting the password from 3rd party credentials.
+    MONGODB_PASSWORD_KEY = "mongodb://" + USER + "@" + HOSTNAME + ":" + str(PORT)
+    print(MONGODB_PASSWORD_KEY)
+    if credentials.get(MONGODB_PASSWORD_KEY):
+        PASSWORD=credentials.get(MONGODB_PASSWORD_KEY)
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MongoDB.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 9]
```diff
@@ -92,21 +92,21 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
 if variables.get("MYSQL_HOSTNAME"):
     HOSTNAME = variables.get("MYSQL_HOSTNAME")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 10]
```diff
@@ -92,21 +92,21 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
...
 else:
-    print("MYSQL_HOSTNAME not defined by the user.")
+    print("The MYSQL_HOSTNAME variable is not provided by the user.")
     sys.exit(1)
 if variables.get("MYSQL_PORT"):
     PORT = int(variables.get("MYSQL_PORT"))
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 11]
```diff
@@ -92,21 +92,21 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
...
 else:
-    print("MYSQL_PORT not defined by the user. Using the default value:", PORT)
+    print("The MYSQL_PORT variable is not provided by the user. Using the default value:", PORT)
 if variables.get("MYSQL_DATABASE"):
     DATABASE = variables.get("MYSQL_DATABASE")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 12]
```diff
@@ -92,21 +92,21 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
...
 else:
-    print("MYSQL_DATABASE not defined by the user.")
+    print("The MYSQL_DATABASE variable is not provided by the user.")
     sys.exit(1)
 if variables.get("MYSQL_USER"):
     MYSQL_USER = variables.get("MYSQL_USER")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 13]
```diff
@@ -276,16 +276,16 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
 if variables.get("MYSQL_HOSTNAME"):
     HOSTNAME = variables.get("MYSQL_HOSTNAME")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 14]
```diff
@@ -276,16 +276,16 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
...
 else:
-    print("MYSQL_HOSTNAME not defined by the user.")
+    print("The MYSQL_HOSTNAME variable is not provided by the user.")
     sys.exit(1)
 if variables.get("MYSQL_PORT"):
     PORT = int(variables.get("MYSQL_PORT"))
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 15]
```diff
@@ -276,16 +276,16 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
...
 else:
-    print("MYSQL_PORT not defined by the user. Using the default value:", PORT)
+    print("The MYSQL_PORT variable is not provided by the user. Using the default value:", PORT)
 if variables.get("MYSQL_DATABASE"):
     DATABASE = variables.get("MYSQL_DATABASE")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DataConnectors/resources/catalog/MySQL.xml [hunk 16]
```diff
@@ -276,16 +276,16 @@ MYSQL_URL_KEY = "mysql://<username>@<hostname>:<port>"
...
 else:
-    print("MYSQL_DATABASE not defined by the user.")
+    print("The MYSQL_DATABASE variable is not provided by the user.")
     sys.exit(1)
 if variables.get("MYSQL_USER"):
     MYSQL_USER = variables.get("MYSQL_USER")
```
Comment (mbenguig) [inline DataConnectors/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: DatabaseServices/METADATA.json [hunk 17]
```diff
@@ -7,14 +7,18 @@
...
+      "NoSQL"
     ],
     "version": "1.0"
   },
   "dependencies": ["CloudAutomationClient"],
   "dataspace" : {
     "target" : "global",
-    "files" : ["resources/dataspace/pima-indians-diabetes.csv"]
+    "files" : ["resources/dataspace/pima-indians-diabetes.csv"],
+    "files" : ["resources/dataspace/star_wars_people.json"]
```
Comment (mbenguig) [inline DatabaseServices/METADATA.json]: "files" : ["resources/dataspace/pima-indians-diabetes.csv", 
             "resources/dataspace/star_wars_people.json"]

File: DatabaseServices/resources/catalog/MongoDB_Database_Interaction.xml [hunk 18]
```diff
@@ -0,0 +1,537 @@
...
+  // Create synchro channel
+  channel = "Service_Instance_" + service_instance_id
+  println("channel: " + channel)
+  synchronizationapi.createChannelIfAbsent(channel, false)
+  synchronizationapi.waitUntil(channel, "RUNNING", "{k,x -> x == true}")
+  
+  // Acquire service endpoint
+  service_instance_data = service_instance_rest_api.getServiceInstanceUsingGET(service_instance_id)
+  def instance_id = service_instance_data.getInstanceId()
+  endpoint = service_instance_data.getInstanceEndpoints().entrySet().iterator().next().getValue()
```
Comment (mbenguig) [inline DatabaseServices/resources/catalog/MongoDB_Database_Interaction.xml]: endpoint = service_instance_data.getInstanceEndpoints().get("MY_KEY")   ?

File: proactive-examples.iml [hunk 19]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<module external.linked.project.id="proactive-examples" external.linked.project.path="$MODULE_DIR$" external.root.project.path="$MODULE_DIR$" external.system.id="GRADLE" external.system.module.group="org.ow2.proactive" external.system.module.version="7.33.0-SNAPSHOT" type="JAVA_MODULE" version="4">
+<module external.linked.project.id="proactive-examples" external.linked.project.path="$MODULE_DIR$" external.root.project.path="$MODULE_DIR$" external.system.id="GRADLE" external.system.module.group="org.ow2.proactive" external.system.module.version="8.2.0-SNAPSHOT" type="JAVA_MODULE" version="4">
```
Comment (jrochas) [inline proactive-examples.iml]: This iml file should be removed and ignored for good.

File: MySQL/resources/catalog/MySQL.xml [hunk 20]
```diff
@@ -198,11 +204,13 @@ echo "$port" > $CONTAINER"_port"
...
           <code language="groovy">
             <![CDATA[
+println("--- BEGIN POST-SCRIPT Start_MySQL ---")
+
 def PA_SCHEDULER_REST_URL = variables.get('PA_SCHEDULER_REST_URL')
 def hostname = new URL(PA_SCHEDULER_REST_URL).getHost()
 def CONTAINER
-if (variables.get("MYSQL_INSTANCE_NAME") != null)
-  CONTAINER=variables.get("MYSQL_INSTANCE_NAME")
+if (variables.get("MYSQL_INSTANCE_NAME_CL") != null)
```
Comment (mbenguig) [inline MySQL/resources/catalog/MySQL.xml]: please update according to comments @ l.165

File: MongoDB/resources/catalog/MongoDB.xml [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+synchronizationapi.put(channel, "DELETE_LAUNCHED", false) 
+synchronizationapi.put(channel, "DELETE_INSTANCE", false) 
+
+// Loop over service instance status
+def current_status = "RUNNING"
+def container = variables.get("MONGODB_INSTANCE_NAME")
+print("container: " + container)
+
+while(current_status=="RUNNING") {
+  sleep(30000);
```
Comment (mbenguig) [inline MongoDB/resources/catalog/MongoDB.xml]: sleep 30 sec.. ?

File: DatabaseServices/resources/catalog/MongoDB_Database_Interaction.xml (hunk 18 above)
Comment (Aminelouati) [reply DatabaseServices/resources/catalog/MongoDB_Database_Interaction.xml]: getInstanceEndpoints() returns a Map composed of a single element.

**Pull request merged**